llvm-bazel/terminfo.bzl

```diff
@@ -0,0 +1,217 @@
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""Repository rules to configure the terminfo used by LLVM.
+
+Most users should pick one of the explicit rules to configure their use of terminfo
+with LLVM:
+- `llvm_terminfo_external` will link against an external Bazel terminfo repository.
+- `llvm_terminfo_system` will link against the system terminfo (non-hermetically).
+- 'llvm_terminfo_disable` will disable terminfo completely.
+
+If you would like to make your build configurable, you can use
+`llvm_terminfo_from_env`. By default, this will disable terminfo, but will inspect
+the environment variable (most easily set with a `--repo_env` flag to the
+Bazel invocation) `BAZEL_LLVM_ZLIB_STRATEGY`. If it is set to `external`,
+then it will behave the same as `llvm_terminfo_external`. If it is set to
+`system` then it will behave the same as `llvm_terminfo_system`. Any other
+setting will disable terminfo the same as not setting it at all.
+"""
+
+#def _llvm_terminfo_external_impl(repository_ctx):
+#    repository_ctx.template(
+#        "BUILD",
+#        repository_ctx.attr._external_build_template,
+#        substitutions = {
+#            "@external_terminfo_repo//:terminfo_rule": str(repository_ctx.attr.external_terminfo),
+#        },
+#        executable = False,
+#    )
+#
+#llvm_terminfo_external = repository_rule(
+#    implementation = _llvm_terminfo_external_impl,
+#    local = True,
+#    attrs = {
+#        "_external_build_template": attr.label(
+#            default = Label("//:terminfo_external.BUILD"),
+#            allow_single_file = True,
+#        ),
+#        "external_terminfo": attr.label(
+#            doc = "The dependency that should be used for the external terminfo library.",
+#            mandatory = True,
+#        ),
+#    },
+#)
+
+def _find_c_compiler(repository_ctx):
+    """Returns the path to a plausible C compiler.
+
+    This routine will only reliably work on roughly POSIX-y systems as it
+    ultimately falls back on the `cc` binary. Fortunately, the thing we are
+    trying to use it for (detecting if a trivial source file can compile and
+    link against a particular library) requires very little.
+    """
+    cc = repository_ctx.which(repository_ctx.os.environ.get("CC"))
+    if not cc:
+        cc = repository_ctx.which("clang")
+    if not cc:
+        cc = repository_ctx.which("gcc")
+    if not cc:
+        cc = repository_ctx.which("cc")
+    return cc
```

Comment:
I feel like it would be pretty reasonable to just require that CC be set and error otherwise. Less magic and something that needs to be set anyway.

Reply:
Unfortunately, `CC` is not typically set. =[ The fact that this project uses it is somewhat specific. Other projects use a custom C++ toolchain configuration rather than any `CC` environment variable, and on typical Linux systems `CC` is not set. So unless Bazel itself *introduces* the environment variable for us to depend on in this way, I think we need some fallback.

The SUS/POSIX "way" is `c99`, `c89`, and `cc` binaries in the `PATH`. =/ I should add `c99` and `c89`. I'm pretty confident that this kind of fallback list is roughly what you want in terms of maximum portability across vaguely Unix-like systems (spanning Linux to macOS).